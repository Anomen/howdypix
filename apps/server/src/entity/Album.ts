import { Column, ViewEntity } from "typeorm";

@ViewEntity({
  expression: `
SELECT
   P.DIR,
   P.SOURCE,
   P.PARENTDIR,
   COUNT(DISTINCT P2.ID) AS NBPHOTOS,
   COUNT(DISTINCT P3.DIR) AS NBALBUMS,
   (
      SELECT
         FILE 
      FROM
         PHOTO P4 
      WHERE
         P4.DIR = P.DIR 
         AND P4.SOURCE = P.SOURCE LIMIT 0,
         1
   )
   AS PREVIEW 
FROM
   PHOTO P 
   LEFT JOIN
      PHOTO P2 
      ON (P2.DIR = P.DIR 
      AND P2.SOURCE = P.SOURCE) 
   LEFT JOIN
      PHOTO P3 
      ON (P3.PARENTDIR = P.DIR 
      AND P3.SOURCE = P.SOURCE) 
GROUP BY
   P.DIR,
   P.SOURCE,
   P.PARENTDIR;
    `,
})
export class Album {
  @Column("text")
  dir: string;

  @Column("text")
  source: string;

  @Column("text")
  parentDir: string;

  @Column("text")
  preview: string;

  @Column("int")
  nbPhotos: number;

  @Column("int")
  nbAlbums: number;
}
