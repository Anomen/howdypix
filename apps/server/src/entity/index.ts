export * from "./Source";
export * from "./Album";
export * from "./Photo";
export * from "./Search";
export * from "./SearchResult";
