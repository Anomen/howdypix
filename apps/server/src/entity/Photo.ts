import {
  Column,
  Entity,
  ManyToOne,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { SearchResult } from "./SearchResult";
import { Album } from "./Album";

export enum PHOTO_STATUS {
  NOT_PROCESSED = "not_processed",
  PROCESSED = "processed",
}

@Entity()
@Unique(["album", "file"])
export class Photo {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  status: PHOTO_STATUS;

  @Column("int")
  inode: number;

  @Column("int")
  mtime: number;

  @Column("int")
  ctime: number;

  @Column("int")
  birthtime: number;

  @Column("int")
  size: number;

  @Column("text", { nullable: true })
  make: string;

  @Column("text", { nullable: true })
  model: string;

  @Column("int", { nullable: true })
  ISO: number;

  @Column("int", { nullable: true })
  shutter: number;

  @Column("int", { nullable: true })
  processedShutter: number;

  @Column("int", { nullable: true })
  aperture: number;

  @Column("int", { nullable: true })
  processedAperture: number;

  @Column("int", { nullable: true })
  createDate: number;

  // TODO to remove
  @Column("text")
  source: string;

  // TODO to remove
  @Column("text")
  parentDir: string;

  // TODO to remove
  @Column("text")
  dir: string;

  @ManyToOne(() => Album, (album) => album.photos)
  public album!: Album;

  @Column("text")
  file: string;

  @OneToMany(() => SearchResult, (searchResult) => searchResult.photo)
  public searchResults!: SearchResult[];
}
