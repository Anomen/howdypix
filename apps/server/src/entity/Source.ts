import {
  Column,
  Entity,
  getConnection,
  Not,
  OneToMany,
  PrimaryGeneratedColumn,
  Unique,
} from "typeorm";
import { Photo as EntityPhoto } from "./Photo";
import { Album } from "./Album";

@Entity()
@Unique(["source"])
export class Source {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("text")
  source: string;

  @Column("text")
  dir: string;

  @OneToMany(() => Album, (album) => album.sourceLk)
  public albums!: Album[];

  async getNbPhotos(): Promise<number> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const { count } = await photoRepository
      .createQueryBuilder()
      .select("COUNT(distinct file)", "count")
      .where({
        source: this.source,
        dir: "",
        file: Not(""),
      })
      .getRawOne();

    return count;
  }

  async getNbAlbums(): Promise<number> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const { count } = await photoRepository
      .createQueryBuilder()
      .select("COUNT(distinct dir)", "count")
      .where({
        source: this.source,
        dir: Not(""),
      })
      .getRawOne();

    return count;
  }

  async getPreview(): Promise<string> {
    const photoRepository = getConnection().getRepository(EntityPhoto);

    const data = await photoRepository
      .createQueryBuilder()
      .select("file", "preview")
      .where({
        source: this.source,
        file: Not(""),
      })
      .getRawOne();

    // TODO change for undefined picture
    return data?.preview || "TOTO.JPG";
  }

  static async upsert(sourceName: string, dir: string): Promise<void> {
    const sourceRepository = getConnection().getRepository(Source);
    const sourceToUpdate = await sourceRepository.findOne({
      source: sourceName,
    });

    if (sourceToUpdate) {
      // Update the value
      sourceToUpdate.dir = dir;
      await sourceRepository.save(sourceToUpdate);
    } else {
      const source = new Source();
      source.source = sourceName;
      source.dir = dir;
      await sourceRepository.save(source);
    }
  }
}
