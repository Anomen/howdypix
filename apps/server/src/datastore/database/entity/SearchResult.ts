import {
  Column,
  Entity,
  PrimaryGeneratedColumn,
  ManyToOne,
  CreateDateColumn,
  UpdateDateColumn,
} from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @CreateDateColumn()
  createdAt: Date;

  @UpdateDateColumn()
  updatedAt: Date;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(() => Search, (search) => search.searchResults)
  public search!: Search;

  @ManyToOne(() => Photo, (photo) => photo.searchResults)
  public photo!: Photo;
}
