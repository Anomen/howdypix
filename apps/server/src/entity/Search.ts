import { Column, Entity, PrimaryGeneratedColumn, OneToMany } from "typeorm";
import { SearchResult } from "./SearchResult";

@Entity()
export class Search {
  @PrimaryGeneratedColumn()
  id: number;

  @Column({ type: "text", nullable: true })
  orderBy: string;

  @Column({ type: "text", nullable: true })
  filterBy: string;

  @Column({ type: "text", nullable: true })
  album: string;

  @Column({ type: "text", nullable: true })
  source: string;

  @OneToMany(
    () => SearchResult,
    searchResult => searchResult.search
  )
  public searchResults!: SearchResult[];
}
