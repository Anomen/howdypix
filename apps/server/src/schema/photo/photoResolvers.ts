import {
  NexusGenArgTypes,
  NexusGenFieldTypes,
  NexusGenRootTypes,
} from "@howdypix/graphql-schema/schema.d";
import { appDebug, generateThumbnailUrls } from "@howdypix/utils";
import { Photo as EntityPhoto } from "../../entity/Photo";
import { ApolloContext } from "../../types.d";
import { appConfig } from "../../config";
import { photoHelpers } from "../../helpers/photoHelpers";

const debug = appDebug("gql");

export const getPhotoResolver = () => async (
  root: {},
  args: NexusGenArgTypes["Query"]["getPhoto"],
  ctx: ApolloContext
): Promise<NexusGenFieldTypes["Query"]["getPhoto"]> => {
  debug(`Fetching photo ${args.file}.`);

  const photoRepository = ctx.connection.getRepository(EntityPhoto);
  const photo = await photoRepository.findOne({
    where: { dir: args.album, source: args.source, file: args.file },
  });

  if (photo) {
    const photoSteam = await photoHelpers.fetchPhotoSteam(
      ctx.connection,
      photo,
      args
    );
    const photoIndexInStream = photoSteam.findIndex(
      (sr) => sr.photoId === photo.id
    );

    return {
      id: photo.id.toString(),
      files: generateThumbnailUrls(appConfig.webapp.baseUrl, {
        file: photo.file,
        source: photo.source,
        name: photo.file,
        dir: photo.dir,
      }).map((tn) => tn.url),
      aperture: photo.processedAperture,
      shutter: photo.processedShutter,
      iso: photo.ISO,
      make: photo.make,
      model: photo.model,
      birthtime: photo.birthtime,
      next: photoSteam[photoIndexInStream + 1]?.photo.file ?? null,
      previous: photoSteam[photoIndexInStream - 1]?.photo.file ?? null,
      photoStream: photoSteam.map(
        (sr): NexusGenRootTypes["PhotoStreamThumbnail"] => ({
          id: String(sr.photo.id),
          file: String(sr.photo.file),
          thumbnails: generateThumbnailUrls(
            appConfig.webapp.baseUrl,
            sr.photo
          ).map((tn) => tn.url),
        })
      ),
    };
  }

  return null;
};
