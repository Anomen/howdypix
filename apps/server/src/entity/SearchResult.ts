import { Column, Entity, PrimaryGeneratedColumn, ManyToOne } from "typeorm";
import { Search } from "./Search";
import { Photo } from "./Photo";

@Entity()
export class SearchResult {
  @PrimaryGeneratedColumn()
  id: number;

  @Column("int")
  searchId!: number;

  @Column("int")
  photoId!: number;

  @Column("int")
  public order!: number;

  @ManyToOne(
    () => Search,
    search => search.searchResults
  )
  public search!: Search;

  @ManyToOne(
    () => Photo,
    photo => photo.searchResults
  )
  public photo!: Photo;
}
