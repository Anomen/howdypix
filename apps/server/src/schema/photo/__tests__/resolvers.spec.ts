import { Connection } from "typeorm";
import { getPhotoResolver } from "../photoResolvers";
import { Photo as EntityPhoto } from "../../../entity/Photo";

describe("getPhotoResolver", () => {
  beforeEach(() => {
    jest.resetAllMocks();
  });

  const resolver = getPhotoResolver();

  test("should return the detail of one picture", async () => {
    const connection = ({
      getRepository: () => ({
        findOne: (): EntityPhoto => ({
          file: "file",
          dir: "dir",
          source: "source",
          birthtime: 123,
          createDate: 456,
          ctime: 789,
          id: 123456789,
          inode: 249508340,
          ISO: 348095,
          make: "make",
          model: "model",
          mtime: 124323423,
          parentDir: "parent",
          size: 123123
        })
      })
    } as unknown) as Connection;

    expect(
      await resolver(
        {},
        { file: "file", album: "album", source: "source" },
        { user: null, connection }
      )
    ).toMatchSnapshot();
  });

  test("should return null if no pictures have been found", async () => {
    const connection = ({
      getRepository: () => ({
        findOne: (): null => null
      })
    } as unknown) as Connection;

    expect(
      await resolver(
        {},
        { file: "file", album: "album", source: "source" },
        { user: null, connection }
      )
    ).toMatchSnapshot();
  });
});
