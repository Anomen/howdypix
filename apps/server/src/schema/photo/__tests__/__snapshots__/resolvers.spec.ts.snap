// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getPhotoResolver should return null if no pictures have been found 1`] = `null`;

exports[`getPhotoResolver should return the detail of one picture 1`] = `
Object {
  "files": Array [
    "http://localhost:3004/static/@source:dir/filex200.jpg",
    "http://localhost:3004/static/@source:dir/filex600.jpg",
    "http://localhost:3004/static/@source:dir/filex1280.jpg",
    "http://localhost:3004/static/@source:dir/filex2048.jpg",
  ],
  "id": "123456789",
  "next": "file3",
  "photoStream": Array [
    Object {
      "file": "file1",
      "id": "345",
      "thumbnails": Array [
        "http://localhost:3004/static/@undefined:file1x200.jpg",
        "http://localhost:3004/static/@undefined:file1x600.jpg",
        "http://localhost:3004/static/@undefined:file1x1280.jpg",
        "http://localhost:3004/static/@undefined:file1x2048.jpg",
      ],
    },
    Object {
      "file": "file2",
      "id": "123456789",
      "thumbnails": Array [
        "http://localhost:3004/static/@undefined:file2x200.jpg",
        "http://localhost:3004/static/@undefined:file2x600.jpg",
        "http://localhost:3004/static/@undefined:file2x1280.jpg",
        "http://localhost:3004/static/@undefined:file2x2048.jpg",
      ],
    },
    Object {
      "file": "file3",
      "id": "3453",
      "thumbnails": Array [
        "http://localhost:3004/static/@undefined:file3x200.jpg",
        "http://localhost:3004/static/@undefined:file3x600.jpg",
        "http://localhost:3004/static/@undefined:file3x1280.jpg",
        "http://localhost:3004/static/@undefined:file3x2048.jpg",
      ],
    },
  ],
  "previous": "file1",
}
`;
